File: statistics/TypeScript/softmax.ts
import {Summation} from "./Summation";
function exp(x:number):number{
    return Math.pow(Math.E,x);
}
function softmax([...value]:Array<number>):Array<number>{
    var softmaxValue:Array<number> = [];
    var cache:Array<number> = [];
    for(let x=0;x<value.length;x++){
        cache.push(exp(value[x]));
    }
    var total = Summation(cache);
    for(let x=0;x<value.length;x++){
        softmaxValue.push(exp(value[x])/total)
    }
    return softmaxValue;
}
export {softmax}

File: statistics/python/softmax.py
# -*- coding: utf-8 -*-
# @Author: Admin
# @Date:   2020-01-10 01:34:45
# @Last Modified by:   jingyuexing
# @Last Modified time: 2020-07-19 05:41:21


def exp(x=0):
    E = 2.718281828459045
    return E**x


def softmax(*values):
    softmaxValue = []
    thisValues = []
    total = 0
    for k in values:
        if isinstance(k, (list, tuple)):
            for m in k:
                thisValues.append(m)
    for i in thisValues:
        total = total + exp(i)
    for n in thisValues:
        softmaxValue.append(exp(n) / total)
    return softmaxValue
